include/envoy/buffer/buffer.h

```diff
@@ -32,6 +38,18 @@ class Instance {
    */
   virtual void add(const void* data, uint64_t size) PURE;
 
+  /**
+   * Add externally owned data to the buffer. Nothing is copied.
+   * The memory at <data> must be valid until all data has been read.
+   * <releaseCallback> is called when the memory is no longer needed by the buffer.
+   * @param data supplies a pointer to the externally owned data.
+   * @param releaseCallback a function to be called
+   * @param releaseArg optional argument passed to releaseCallback as <arg>.
+   * @param size supplies the size of the data.
+   */
+  virtual void addReference(const void* data, uint64_t size, refReleaseCb releaseCallback,
```

Comment:
nit: please use snake case for all variables. E.g., `release_callback`, `release_arg`, etc. Same elsewhere.

In general I agree with @ggreenway that if possible we should use a more idiomatic C++ interface vs. C interface, but I'm not sure this is going to be possible without going through a whole bunch of contortions given that we have to store all of this inside evbuffer. There are a bunch of things we could do around storing things inside buffer itself, and adding the evbuffer ugliness.

I would like to have a larger conversation though, which is that this behavior is inherently very risky, and depending on how it is used the releaser very likely has to be fully static, given that buffers can potentially live longer than filters and data gets moved around a lot. 

Out of curiosity where do you plan on using this?  I'm just wondering if it's worth adding this feature on top of evbuffer now or we should wait until we get rid of evbuffer which I want to do anyway.

Reply:
I think that in @genioshelo's case the release is static, since this a global ref counted pool, but will defer to his comment here.

What would an ideal API look like here if we didn't have to rely on libevent and wanted to avoid issues with dangling references? Would weak references help, or are they a design anti-pattern, since they would hide corner cases with memory ownership?

Reply:
@mattklein123 The basic use case we're thinking of is being able to pass in ref-counted chunks in a memory cache to Envoy. The passed in data indeed should not be tied to the filter's lifetime.

Ref waiting for an overhaul of the buffer class: for our specific use case, this depends on the timing. This would likely be the only additional functionality we need for now, so it may be inefficient to hold off for a complete re-write. (Although I do think the Buffer interface could use some reinvention.) Is there a general timeline/need for this?

@htuch I'll add a comment on about the silo thread for the releaser. Honestly I am not really sure about an ideal API - I haven't spent enough time in C++ to have a strong opinion either way. I'd be interested to hear from @mattklein123 how a re-design of buffer would address this.

Reply:
Rewriting the buffer implementation to get rid of evbuffer is not resourced, so I wouldn't consider it a thing that will happen any time soon.

In a perfect world, IMO we would have a C++ interface for a constant buffer fragment, which has `incRef()`, `decRef()`, `data()`, and `len()` on it. Then buffer fragments could be added to the buffer by pointer.

I think it's possible to do this interface and map this onto C evbuffer by creating this interface and passing it in to evbuffer with a static releaser which does a static cast and which just calls `decRef()`. 

@genioshelo can you time box mocking up ^ ? I don't think it would be that hard and would make the interface a lot cleaner which if we converted to native C++ later would be easy to work with.

Reply:
@mattklein123 Le me see if I get your idea. It would be to have a `BufferFragment` class which is created with a destructor/releasor callback (or optionally without if it owns the data) and the `incRef()`, `decRef()`, `data()`, and `len()` methods. Then there would be a method `addBufferFragment()` on the buffer interface to pass a pointer to a `BufferFragment` object. The buffer can internally deal with passing the data pointer to evbuffer.

Does that sound about right?

Reply:
Yup exactly. For evbuffer, you pass the pointer to the fragment as the arg, and a static method to static cast and dec ref the fragement. It should be pretty clean to the caller.

Reply:
Yeah this sounds cleaner to the caller. And I assume the caller of `addBufferFragment` would own the fragment object itself being passed in by reference/pointer.

I can work on modifying this pull request to use this interface instead.

Reply:
Or would you rather I abandon this one and work on a new pull request?

Reply:
You can just work on this one. Whatever is easier.

Reply:
Alright, I'll stick with this one.